### PR TITLE
[lld][AArch64][ELF][PAC] Support `.relr.auth.dyn` section

### LINKED ISSUE #95204: [lld][AArch64][ELF][PAC] Support `.relr.auth.dyn` section
Previously merged PR #87635 introducing `.relr.auth.dyn` section support caused an issue described in https://github.com/llvm/llvm-project/pull/87635#issuecomment-2155581767
The PR was reverted #94843.

Need to fix the issue and re-submit the changes with proper fixes.

### PR DESCRIPTION
**Pull request by kovdan01**
This re-applies #87635 after the issue described in https://github.com/llvm/llvm-project/pull/87635#issuecomment-2155318065 is fixed in ebc123e0793a1cbcb69b4af1548e339e018ffff2. A corresponding test is also added.

Original PR description below.

Support `R_AARCH64_AUTH_RELATIVE` relocation compression as described in https://github.com/ARM-software/abi-aa/blob/main/pauthabielf64/pauthabielf64.rst#relocation-compression.

### CODE REVIEW

**Comment by zeroomega** [comment]
I confirm it no longer generate empty `.rel.dyn` or invalid sh_info on REL section when building an relocatable object file.

**Review by zeroomega — APPROVED** [review]
LGTM from my point of view. I compared it with the original implementation in #87635 and confirm the empty and invalid `.rel.dyn` section in relocatable object file was fixed. 
Please wait for @MaskRay's approval before landing it.

**Review by MaskRay — CHANGES_REQUESTED** [review]
`.rela.dyn` is currently created outside of the `config->hasDynSymTab` condition.
In relocatable links, `.rela.dyn` will be discarded by `removeUnusedSyntheticSections`.
The reverted `.relr.auth.dyn` patch incorrectly retained the empty `.relr.dyn`.

I think we should suppress the creation of `.rela.dyn` in the first place.
I will try updating `createSyntheticSections`.

The `!config->relocatable` condition in `removeUnusedSyntheticSections` in this PR can then be dropped.

File: lld/test/ELF/aarch64-reloc-pauth.s
Comment (MaskRay) [inline lld/test/ELF/aarch64-reloc-pauth.s:183]: `-o empty-rela.ro`

**Comment by kovdan01** [comment]
> I think we should suppress the creation of `.rela.dyn` in the first place.
> I will try updating `createSyntheticSections`.

@MaskRay I agree that it's better not to construct the section at all if we know it's not needed than construct an empty one and then delete it, thanks for implementing that it ebc123e0793a1cbcb69b4af1548e339e018ffff2.

> The `!config->relocatable` condition in `removeUnusedSyntheticSections` in this PR can then be dropped.

Deleted, thanks. I also updated the PR description.

File: lld/test/ELF/aarch64-reloc-pauth.s
Comment (kovdan01) [reply lld/test/ELF/aarch64-reloc-pauth.s]: Renamed, see 87d2e3f1894a3296a04f79a69385daabbc92884e

**Review by MaskRay — APPROVED** [review]

**Comment by kovdan01** [comment]
Pushed a merge commit to re-trigger buildkite job which failed to unrelated reasons

**Pull request merged**
